mock_test.go

```diff
@@ -210,3 +211,107 @@ func TestPeek(t *testing.T) {
 		t.Fatal("expected Peek() to return false")
 	}
 }
+
+// TestTickerFunc_ContextDoneDuringTick tests that TickerFunc.Wait() can't return while the tick
+// function callback is in progress.
+func TestTickerFunc_ContextDoneDuringTick(t *testing.T) {
+	t.Parallel()
+	testCtx, testCancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer testCancel()
+	mClock := quartz.NewMock(t)
+
+	tickStart := make(chan struct{})
+	tickDone := make(chan struct{})
+	ctx, cancel := context.WithCancel(testCtx)
+	defer cancel()
+	tkr := mClock.TickerFunc(ctx, time.Second, func() error {
+		close(tickStart)
+		select {
+		case <-tickDone:
+		case <-testCtx.Done():
+			t.Error("timeout waiting for tickDone")
+		}
+		return nil
+	})
+	w := mClock.Advance(time.Second)
+	select {
+	case <-tickStart:
+		// OK
+	case <-testCtx.Done():
+		t.Fatal("timeout waiting for tickStart")
+	}
+	waitErr := make(chan error, 1)
+	go func() {
+		waitErr <- tkr.Wait()
+	}()
+	cancel()
+	select {
+	case <-waitErr:
+		t.Fatal("wait should not return while tick callback in progress")
+	case <-time.After(time.Millisecond * 100):
+		// OK
+	}
+	close(tickDone)
+	select {
+	case err := <-waitErr:
+		if !errors.Is(err, context.Canceled) {
+			t.Fatal("expected context.Canceled error")
+		}
+	case <-testCtx.Done():
+		t.Fatal("timed out waiting for wait to finish")
+	}
+	w.MustWait(testCtx)
+}
+
+// TestTickerFunc_LongCallback tests that we don't call the ticker func a second time while the
+// first is still executing.
+func TestTickerFunc_LongCallback(t *testing.T) {
+	t.Parallel()
+	testCtx, testCancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer testCancel()
+	mClock := quartz.NewMock(t)
+
+	expectedErr := errors.New("callback error")
+	tickStart := make(chan struct{})
+	tickDone := make(chan struct{})
+	ctx, cancel := context.WithCancel(testCtx)
+	defer cancel()
+	tkr := mClock.TickerFunc(ctx, time.Second, func() error {
+		close(tickStart)
+		select {
+		case <-tickDone:
+		case <-testCtx.Done():
+			t.Error("timeout waiting for tickDone")
+		}
+		return expectedErr
+	})
+	w := mClock.Advance(time.Second)
+	select {
+	case <-tickStart:
+		// OK
+	case <-testCtx.Done():
+		t.Fatal("timeout waiting for tickStart")
+	}
+	// second tick completes immediately, since it doesn't actually call the
+	// ticker function.
+	mClock.Advance(time.Second).MustWait(testCtx)
+
+	waitErr := make(chan error, 1)
+	go func() {
```

Comment:
We read from `waitErr` on L307, so a waitgroup is redundant. On success, we know this goroutine has terminated.

Generally speaking, we mainly care about whether the _code under test_ leaks goroutines. Leaking a test goroutine almost never matters, since the test is finite (the exception is if the leaked goroutine interferes with other tests, which is rare).